src/omnicore/omnicore.cpp

```diff
@@ -2367,15 +2372,38 @@ bool mastercore::UseEncodingClassC(size_t nDataSize)
     return nTotalSize <= nMaxDatacarrierBytes && fDataEnabled;
 }
 
+/**
+ * Determines whether Class D is active & whether the compressed payload is suitable size.
+ *
+ * @param nDataSize The length of the payload
+ * @return True, if Class D is enabled and the payload is small enough
+ */
+bool mastercore::UseEncodingClassD(size_t nDataSize)
+{
+    size_t nTotalSize = nDataSize + GetOmMarker(OMNI_CLASS_D).size(); // Marker "ol"
+    int nBlockNow = GetHeight();
+    bool fDataEnabled = GetBoolArg("-datacarrier", true);
+    bool fClassDEnabled = IsFeatureActivated(FEATURE_CLASS_D, nBlockNow);
+    if (nDataSize == 0) { // no compressed payload supplied
```

Comment:
This check is to make sure we fall back to Class C if there is no compressed payload available.

More detail:

`WalletTxBuilder()` has parameters for both a regular payload and a compressed payload.  

There are certain functions that intentionally don't use compressed payloads (eg management functions like deactivation & alerting don't use Class D yet just in case we find a bug in it) and things like raw transaction creation also aren't Class D aware.

In these cases the compressed payload parameter passed to `WalletTxBuilder()` is empty, which means this check will cause `UseEncodingClassD()` to return false (and thus cause a fallback to Class C).

Hope that makes sense :)

I could move the check for an empty compressed payload into `WalletTxBuilder()` itself rather than `UseEncodingClassD()` if you think that is a more suitable place for it?